subsys/net/ip/net_context.c

```diff
@@ -301,6 +304,26 @@ static void handle_ack_timeout(struct k_work *work)
 		net_context_unref(tcp->context);
 	}
 }
+
+static void handle_timewait_timeout(struct k_work *work)
+{
+	struct net_tcp *tcp = CONTAINER_OF(work, struct net_tcp,
+					   timewait_timer);
+
+	NET_DBG("Timewait expired in %dms", TIMEWAIT_TIMEOUT);
+
+	if (net_tcp_get_state(tcp) == NET_TCP_TIME_WAIT) {
+		net_tcp_change_state(tcp, NET_TCP_CLOSED);
+
+		if (tcp->context->recv_cb) {
+			tcp->context->recv_cb(tcp->context, NULL, 0,
```

Comment:
@rveerama1 : Any reason why you put this code here? All the states like FIN_WAIT, TIME_WAIT are internal TCP states. From the application point of view, the connection is closed when first (valid) FIN is received. The rest is again just the internal TCP connection shutdown sequence. So, calling recv_cb(NULL) at the end of TIME_WAIT expiration seems like too late. I'm considering submitting a patch to remove it from here, unless you have better suggestions. Thanks.

Reply:
I wasn't sure what kind of corner cases hits this original code without the timer. If you have any cases which is causing issues with it, please send a fix.